dyno: Begin migrating production parser checks to uAST builder

dlongnecke-cray opened the pull request:
dyno: Begin migrating production parser checks to uAST builder (#19916)

Add a pass over freshly parsed uAST trees to the builder that is
equivalent to the production compiler pass `checkParsed`.

Some, but not all of the checks are present here. More will be added
as part of future work.

Add a `CompilerFlags` class intended to store binary compiler flags.
Not every flag present in a invocation of `chpl --help` will be
stored here, but it is a good place to place binary flags that can
be set once per revision (such as `--warn-unstable`).

Adjust the production compiler's error handler to print out headers
(such as `In function 'foo'`) in a fashion similar to the production
compiler. This is not very sophisticated and does not do fancy
things like print out the callstack. That should be tackled by
code internal to dyno as part of future work.

Adjust the `parseFileToBuilderResult` query so that it no longer
reports any errors to the context. The `parse` family of queries
(which use the `parseFileTo...` query) do report the errors. The
distinction is mentioned in the docs. This was needed to prevent
custom error handlers from naively triggering a query cycle.

TESTING

- [x] `ALL` on `linux64`, `standard`
- [x] `ALL` on `linux64`, `COMM=gasnet`

Reviewed by @arezaii, @mppf. Thanks!

Signed-off-by: David Longnecker <dlongnecke-cray@users.noreply.github.com>

dlongnecke-cray commented:
The CI checks are failing due to some "static method never called" kinds of errors. I'd like to keep filling out the rest of the checks before I merge this, anyhow.

dlongnecke-cray commented:
@mppf Does this general approach seem acceptable to you? I'm leaning a bit on maintaining a stack of AstNode pointers instead of using IDs, because I figure that's OK (as this is all happening inside a single query).

---

My comments:

I worry that there is some duplication of "boilerplate" visitor code here in that I've defined methods to get the parent and search upwards through parents. I wonder if this will be an issue down the line?

mppf commented:
> @mppf Does this general approach seem acceptable to you?

Yes. My only suggestion would be that, if the checking needs to traverse all of the parsed uAST, it should be more memory efficient if we can do the checking and the ID assignment in a single traversal. At the moment it's not really clear to me if we could just have the ID assignment call some `visit` function on the checker. Either way I am happy having the checking be in a separate file from the builder.

> I'm leaning a bit on maintaining a stack of AstNode pointers instead of using IDs, because I figure that's OK (as this is all happening inside a single query).

Yes that is fine an reasonable in this case.

> I worry that there is some duplication of "boilerplate" visitor code here in that I've defined methods to get the parent and search upwards through parents. I wonder if this will be an issue down the line?

The Builder has maps about how to go from a node to its parents which you could be using, alternatively. (Later on, the "get parent ID" query would be the thing to use in most cases). Anyway I'm not quite sure what your concern is here - are you saying this is a common thing for visitors to want to do and so we should roll it into some visitor helper class?

arezaii commented on compiler/dyno/include/chpl/queries/session-queries.h:
```diff
@@ -0,0 +1,47 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
```
Might be nice to have some doc guidance on what kind of queries we expect here, or the context that explains the meaning of a "session".

mppf replied:
Other terms might be "compiler invocation" or maybe "compiler configuration". I don't have any strong preference on the name right now but I think I slightly prefer something more along the lines of "compilation configuration". We have some other stuff in parsing-queries.h that arguably would belong in a "compiler configuration" area -- setModuleSearchPath / setBundledModulePath and related things.

dlongnecke-cray replied:
Ok, is it fine for this header to live in `chpl/queries`? Do you want `-queries.h` at the end of it to follow convention? Or just `compiler-configuration.h`?

mppf replied:
I have been thinking maybe it's not a good practice to have `bla-queries.h` since it's just a pile of functionality and whether it is a query or not is up to individual functions. I'm OK with putting it in `chpl/queries` but arguably it is off-topic there. OK for now though. Maybe `chpl/compiler` (including a new namespace) or something would be called for.

arezaii reviewed (approved):
LGTM

dlongnecke-cray commented on compiler/dyno/test/uast/testStringify.cpp:
```diff
@@ -310,10 +310,6 @@ static void test1(Parser* parser) {
                while true do
                  ;
-               proc noOp(x:int);
-               {
```
It is only allowed when the procedure is extern, I believe.

arezaii replied:
I see that the comment didn't quite line up where I wanted it. I was wondering if the we don't support the empty block itself, outside of the empty statement `;` that `noOp` previously defined.

```chapel
{
  // this block is empty
}
```

dlongnecke-cray replied:
Oops, I think you're right that I deleted a bit too much stuff :D

mppf commented on compiler/dyno/include/chpl/queries/session-queries.h:
```diff
@@ -0,0 +1,47 @@
+  /**
+    Returns true if a given flag is set.
+  */
+  bool isFlagSet(Context* context, Flags::Name flag);
```
`hasFlag` ?

mppf replied:
oh maybe not since it's about the current session

mppf commented on compiler/dyno/include/chpl/queries/session-queries.h:
```diff
@@ -0,0 +1,47 @@
+  void setFlagsList(Context* context, Flags list);
+
+  /**
+    Returns true if a given flag is set.
```
"in the current session"

mppf commented on compiler/dyno/include/chpl/queries/Flags.h:
```diff
@@ -0,0 +1,94 @@
+  class is intended to be a convenient place to store various different
+  boolean flags the compiler may need to check.
+*/
+class Flags final {
```
IMO this one would be better named CompilationFlags or something like that. I feel "Flags" is too general (I was expecting it to be a bit vector, with that name).

dlongnecke-cray replied:
I agree, I'm fine with renaming it CompilationFlags as well.

dlongnecke-cray commented:
I've renamed `Flags` to `CompilerFlags` and I've renamed the header to `compiler-configuration.h`. It still lives in `queries` (soon to be `framework`), and honestly with the rename to `framework` I don't really feel bad about leaving the header there, but others might disagree.

mppf commented:
> I've renamed `Flags` to `CompilerFlags` and I've renamed the header to `compiler-configuration.h`. It still lives in `queries` (soon to be `framework`), and honestly with the rename to `framework` I don't really feel bad about leaving the header there, but others might disagree.

Let's leave it there for now but if we find we need to add more stuff related to how a Chapel compiler invocation works we might add a new directory for that.
